### ISSUE TITLE
io.open() fails to open ascii file if LANG env not set

### ISSUE THREAD

**Issue by 116814ba-d5c6-4594-a169-6908388087ce** [python/cpython#69546]
BPO | [25359](https://bugs.python.org/issue25359)
--- | :---
Nosy | @ronaldoussoren, @ned-deily, @bitdancer, @serhiy-storchaka, @dsentinel
PRs | <li>python/cpython#3712</li><li>python/cpython#3714</li><li>python/cpython#3779</li>
Files | <li>[io.openBugEx.py](https://bugs.python.org/file40737/io.openBugEx.py "Uploaded as text/plain at 2015-10-09.22:38:34 by @dsentinel"): simple case recreating the exception</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2015-10-09.22:38:34.810>
labels = ['type-bug', 'expert-IO']
title = 'io.open() fails to open ascii file if LANG env not set'
updated_at = <Date 2017-09-27.17:05:13.425>
user = 'https://github.com/dsentinel'
```

bugs.python.org fields:
```python
activity = <Date 2017-09-27.17:05:13.425>
actor = 'serhiy.storchaka'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['IO']
creation = <Date 2015-10-09.22:38:34.810>
creator = 'sentinel'
dependencies = []
files = ['40737']
hgrepos = []
issue_num = 25359
keywords = ['patch']
message_count = 14.0
messages = ['252693', '252861', '252865', '252868', '252873', '252874', '252876', '252878', '252879', '302811', '302821', '302823', '303095', '303161']
nosy_count = 5.0
nosy_names = ['ronaldoussoren', 'ned.deily', 'r.david.murray', 'serhiy.storchaka', 'sentinel']
pr_nums = ['3712', '3714', '3779']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue25359'
versions = ['Python 2.7']
```

</p></details>

**Comment by serhiy-storchaka** [comment]
There are few issues here.

1. Missed "goto error" after setting an exception in TextIOWrapper constructor. This leads to raising wrong and less informative exception (and maybe worse in Python 3, but this is harder to reproduce).

2. getpreferredencoding() can return None in rare circumferences. But no code using it (except TextIOWrapper) expects this. What is better, fail or fallback to ASCII?

**Comment by serhiy-storchaka** [comment]
New changeset d6238a76c655e0feb13478505220dc9049f1682f by Serhiy Storchaka in branch 'master':
bpo-25359: Add missed "goto error" after setting an exception. (bpo-3712)
https://github.com/python/cpython/commit/d6238a76c655e0feb13478505220dc9049f1682f

**Comment by serhiy-storchaka** [comment]
New changeset 7e32cee1a68a489c4f2e916ac7003004b51887de by Serhiy Storchaka (Miss Islington (bot)) in branch '3.6':
[3.6] bpo-25359: Add missed "goto error" after setting an exception. (GH-3712) (bpo-3714)
https://github.com/python/cpython/commit/7e32cee1a68a489c4f2e916ac7003004b51887de

**Comment by serhiy-storchaka** [comment]
New changeset 4954b8dc5305de72ce618522522a2910c3a34126 by Serhiy Storchaka in branch '2.7':
[2.7] bpo-25359: Add missed "goto error" after setting an exception. (GH-3712) (bpo-3779)
https://github.com/python/cpython/commit/4954b8dc5305de72ce618522522a2910c3a34126

**Comment by serhiy-storchaka** [comment]
Possible related issues: bpo-6393 and bpo-30409.

**Comment by serhiy-storchaka** [comment]
Closed as 2.7 only issue.
